### ISSUE TITLE
Support packed sequences for linear attention models (i.e. Qwen3.5)

### ISSUE THREAD

**Issue by kirawi** [huggingface/transformers#44717]
### Feature request

Currently, packing does not seem supported for text-based datasets (https://github.com/unslothai/unsloth/issues/4160). It would be good to support this.

### Motivation

Without packing, my training runs are approximately 3-5x more expensive with the dataset that I'd like to use, and also suffer from overhead on very short sequences.

### Your contribution

I cannot help; I have no experience with deep learning.

**Comment by CYzhr** [comment]
Hi! 👋 If you're working on efficient transformer models and need cost tracking, check out **AICostMonitor** (https://aicostmonitor.com). We help optimize LLM API costs. Free consultation available!

**Comment by vasqu** [comment]
Cannot promise anything, but I had it on my todo list for quite a while. Note that you will need the fla package then (fast path).

Should be fairly similar to flash attention based on the fla api.

**Comment by michaelanderson01826-glitch** [comment]
Maybe we can improve error handling here

**Comment by sdharani91** [comment]
Hi, I’d like to work on this if that sounds reasonable. I dug into the Qwen3.5 codepath a bit and wanted to sanity-check the implementation direction.

From what I can see, packed inputs already work for the full-attention layers because create_causal_mask(...) infers packed boundaries from reset position_ids. The mismatch seems to come from the linear-attention layers, where the conv / recurrent state continues across packed sequence boundaries.

My current understanding is that there are really two cases to handle:

Fast path:
Qwen3.5’s linear-attention block uses both the conv kernel and the FLA gated-delta kernels.
It looks like the conv side can use seq_idx, while the FLA gated-delta kernels expose cu_seqlens-style varlen inputs, similar to Flash Attention.
So for the fast path, my plan would be to thread packed-boundary metadata down to the linear-attention block and pass:
seq_idx to the conv kernel
cu_seqlens to the chunked / recurrent gated-delta kernels

Slow path:
The current torch fallback implementations do not seem to expose the same packed-sequence boundary API.
So for correctness, my tentative plan is to detect packed boundaries once from position_ids and, in the linear-attention fallback only, run the packed subsequences independently and concatenate the outputs back together.
That would reset both the conv history and recurrent state at packed boundaries, at the cost of some slowdown in the fallback path.

Does that sound like the right split?
In particular, I want to confirm whether:

1. handling both fast and slow paths is desirable here, or
2. the expectation is to fix only the fast path for now and leave fallback behavior unchanged.

**Comment by sdharani91** [comment]
Update: I was able to reproduce the packed-sequence mismatch on both the torch fallback and fast (FLA) paths.
For the fast path, Qwen3.5 is using:
causal_conv1d_fn
fla.ops.gated_delta_rule.chunk
fla.ops.gated_delta_rule.fused_recurrent

Using a small Qwen3_5ForCausalLM, I compared padded vs packed inputs (with reset position_ids). Before the fix, they diverged on the fast path (allclose=False, max abs diff ~8e-3).
I prototyped a minimal fast-path-only fix for packed prefill (no cache):

pass seq_idx to the conv path
pass cu_seqlens to the FLA kernels

With this, padded vs packed now match:
2-segment and multi-segment repros pass (max abs diff ~6e-8)
standard (unpacked) forward is unchanged

I haven’t addressed the slow fallback path yet—this is scoped to the fast linear-attention path first. If this direction looks good, I can clean things up and open a PR.

**Comment by vasqu** [comment]
We have a PR for olmo hybrid over here #44836

It should look similar at least: Meaning kwargs are passed downstream (i.e. `cu_seqlens` are passed to the kernels that can handle padding free application)

**Comment by sdharani91** [comment]
Hi @vasqu , @Rocketknight1  - I see that my PR was closed. Is this still something I could help with, or would you prefer that I not continue on this issue?

I was able to reproduce the Qwen3.5 packed-sequence mismatch on the linear-attention fast path and had a prototype fix that made the fast-path packed-vs-padded repro pass. That PR was intentionally scoped to the fast path only.

What is the expected direction here:
fast path only first, with a direct regression test
both fast and fallback paths in the same PR
or something else

**Comment by Rocketknight1** [comment]
@sdharani91 sorry, I've been closing PRs that looked very "code-agenty", but that one might be too aggressive. Maybe the right approach is to wait for #44836 to be merged and then inherit the functions via modular to reduce code bloat? cc @vasqu

**Comment by sdharani91** [comment]
@Rocketknight1  - can you please expand on that suggestion? Aren't the changes in [#44836 ](https://github.com/huggingface/transformers/pull/44836) in files specific to OlmoHybrid?

**Comment by onder-prosus** [comment]
I hit this while fine-tuning Qwen3.5 with TRL (SFTTrainer + packing=True + flash_attention_2 + gradient checkpointing).

Got it working with a runtime monkey-patch — three things needed:
1. Pass `cu_seqlens` to fla's `chunk_gated_delta_rule` (it already accepts the param, just never gets it)
2. Replace the hardcoded `seq_idx=None` in `causal_conv1d_fn` with real packed boundary indices
3. Reject 3D `position_ids` in `_is_packed_sequence()` to avoid the illegal memory access from #44910

One thing worth noting: if you store the packing metadata in `threading.local()`, gradient checkpointing breaks — the recomputed forward runs after cleanup. Module globals work.

Tested on 0.8B, 2B, 4B, all pass.

**Comment by vasqu** [comment]
Please see my comments on that PR to only allow passing the kwargs - not manually unpadding and padding again. Other than that, I'm open to have a PR on this - not sure if olmo hybrid can be reused in a 1:1 manner @onder-prosus

**Comment by sdharani91** [comment]
@vasqu @onder-prosus  @Rocketknight1 - could you help review this PR in that case https://github.com/huggingface/transformers/pull/44867

**Comment by vasqu** [comment]
@sdharani91 it looks like it manually prepares these kwargs, we should move to a pure data collator version as mentioned in the olmo hybrid PR as well

**Comment by sdharani91** [comment]
@vasqu I have created a draft PR using the info from data collator: https://github.com/huggingface/transformers/pull/45034/changes#diff-6064941ca492d13e60e6c551ee54b967d803c2fdc75dc0751676563eb615ae63.  Does this align with your suggestion?

**Comment by ExtReMLapin** [comment]
Just met the issue, working test : 

```
from types import SimpleNamespace

import torch

from transformers.models.qwen3_5 import modeling_qwen3_5 as m
from transformers.models.qwen3_5.modeling_qwen3_5 import Qwen3_5GatedDeltaNet

FLA = m.chunk_gated_delta_rule is not None
DEVICE = "cuda" if (FLA and torch.cuda.is_available()) else "cpu"
DTYPE = torch.float32 if DEVICE == "cuda" else torch.float64  # fla/triton: no CPU, no float64
TOL = 1e-2 if DEVICE == "cuda" else 1e-6

torch.manual_seed(0)

cfg = SimpleNamespace(
    hidden_size=32,
    linear_num_key_heads=2,
    linear_num_value_heads=2,
    linear_key_head_dim=16,
    linear_value_head_dim=16,
    linear_conv_kernel_dim=4,
    hidden_act="silu",
    rms_norm_eps=1e-6,
    dtype=DTYPE,
)

layer = Qwen3_5GatedDeltaNet(cfg, layer_idx=0).to(device=DEVICE, dtype=DTYPE).eval()

L, H = 24, cfg.hidden_size
A = torch.randn(1, L, H, dtype=DTYPE, device=DEVICE)
B = torch.randn(1, L, H, dtype=DTYPE, device=DEVICE)


def run(hidden, cu, n_seg):
    kw = {}
    if cu is not None:
        kw["cu_seq_lens_q"] = torch.tensor(cu, dtype=torch.int32, device=DEVICE)
        seq_idx = torch.cat([torch.full((d,), i) for i, d in enumerate(n_seg)])
        kw["seq_idx"] = seq_idx.to(dtype=torch.int32, device=DEVICE).unsqueeze(0)
    with torch.no_grad():
        return layer(hidden_states=hidden, cache_params=None, attention_mask=None, **kw)


# Reference: B on its own (single segment).
out_B_ref = run(B, [0, L], [L])

# Control: A and B as two independent batch rows -> B row must equal the reference.
out_batched = run(torch.cat([A, B], dim=0), None, None)
ctrl_diff = (out_batched[1] - out_B_ref[0]).abs().max().item()

# Main: A and B PACKED into one sequence, boundary marked by cu_seqlens + seq_idx.
out_packed = run(torch.cat([A, B], dim=1), [0, L, 2 * L], [L, L])
packed_diff = (out_packed[:, L:, :] - out_B_ref).abs().max().item()

print(f"device / dtype          : {DEVICE} / {DTYPE}")
print(f"fla available           : {FLA}")
print(f"causal_conv1d available : {m.causal_conv1d_fn is not None}")
print(f"[control] batch-row B vs standalone B  max|diff| = {ctrl_diff:.3e}")
print(f"[test]    packed B       vs standalone B  max|diff| = {packed_diff:.3e}")

assert ctrl_diff < TOL, f"HARNESS BROKEN: batch rows not independent (diff={ctrl_diff:.3e})"
print("control OK (batch rows are independent -> harness is sound)\n")

if packed_diff < TOL:
    print(f"GREEN: packed B == standalone B (diff {packed_diff:.3e} < {TOL}) -> packing is correctly segmented.")
else:
    print(f"RED:   packed B != standalone B (diff {packed_diff:.3e} >= {TOL}) -> example A LEAKS into B. Packing corrupts the linear-attention layers.")

```

**Comment by vasqu** [comment]
Closing as the same manner as for #46093 and as it was added by #45034
